c-dependencies/js-compute-runtime/builtins/backend.cpp

```diff
@@ -0,0 +1,705 @@
+#include <iostream>
+#include <optional>
+
+// TODO: remove these once the warnings are fixed
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Winvalid-offsetof"
+#include "js/experimental/TypedData.h"
+#pragma clang diagnostic pop
+
+#include "backend.h"
+#include "host_call.h"
+#include "js/Conversions.h"
+
+bool isDigit(char character) { return character >= 48 && character <= 57; }
+bool isLowerAlpha(char character) { return character >= 97 && character <= 122; }
+bool isUpperAlpha(char character) { return character >= 65 && character <= 90; }
+bool isAlphaNumeric(char character) {
+  if (!isDigit(character) && !isLowerAlpha(character) && !isUpperAlpha(character)) {
+    return false;
+  }
+  return true;
+}
+bool isDash(char character) { return character == 45; }
+bool isDot(char character) { return character == 46; }
+
+bool isNotAlphaNumericOrDash(char character) {
+  return !isDigit(character) && !isLowerAlpha(character) && !isUpperAlpha(character) &&
+         !isDash(character);
+}
+
+bool isNotAlphaNumericDotOrDash(char character) {
+  return !isDigit(character) && !isLowerAlpha(character) && !isUpperAlpha(character) &&
+         !isDash(character) && !isDot(character);
+}
+
+bool isValidHost(std::string host) {
+  // ValidHostRegex =
+  // "^(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9\-]*[A-Za-z0-9])(:[0-9]+)$";
+  auto firstCharacter = host.front();
+  // check first character is in the regex [a-zA-Z0-9]
+  if (!isAlphaNumeric(firstCharacter)) {
+    return false;
+  }
+  // split the hostname from the port
+  int pos = host.find_first_of(':');
+  std::string hostname = host.substr(0, pos);
```

Comment:
This will allocate a new string, would it be better to parse a `std::string_view` instead?

Reply:
Yes it would - thank you for spotting this. I've scanned through this file and found a couple more places which could change from string to string_view and have made all the changes in [5fcb9dfdfa86026529ea933260a5c8936d9e922f](https://github.com/fastly/js-compute-runtime/pull/250/commits/5fcb9dfdfa86026529ea933260a5c8936d9e922f)